backend/pkg/auth/oidc.go

```
@@ -132,60 +137,56 @@ func NewOIDCAuthenticator(config *OIDCAuthConfig) Authenticator {
 	return oidcAuthenticator
 }
 
-func (oa *oidcAuth) SetupRouter(router ginhelpers.Router) {
-	router.Use(ginsessions.SessionsMiddleware(oa.sessionStore, "oidc"))
-	oidcRouter := router.Group("/login", "oidc")
-	oidcRouter.GET("/cb", oa.loginCb)
-	oidcRouter.GET("/", oa.login)
-	oidcRouter.GET("/validate_token", oa.validateToken)
+func (oa *oidcAuth) SetupRouter(router *echo.Echo) {
+	router.Use(echosessions.SessionsMiddleware(oa.sessionStore, "oidc"))
 }
 
-func (oa *oidcAuth) validateToken(c *gin.Context) {
-	ctx := c.Request.Context()
+func (oa *oidcAuth) ValidateToken(c echo.Context) error {
+	ctx := c.Request().Context()
 
 	// set request id in response header
-	requestID := c.Writer.Header().Get("X-Request-ID")
+	requestID := c.Response().Writer.Header().Get("X-Request-ID")
 
 	// Check is the id token exists in the request
 	token := tokenFromRequest(c)
 	if token == "" {
 		logger.Debug().Str("request_id", requestID).Msg("ValidateToken, Authorization header is empty")
 		httpError(c, http.StatusUnauthorized)
-		return
+		return nil
 	}
 
 	// If refresh token is not available in the session
 	// mark the request as unauthorized so that the session
 	// can be recreated with refresh_token
-	session := ginsessions.GetSession(c)
+	session := echosessions.GetSession(c)
```

Here is an example of where `GetSession` is not checked for errors. `ginsessions.GetSession(c)` would always return something. But that is not the case for `echosessions.GetSession(c)`, as it can return nil in an error condition.

@yolossn  Should I update all users of `echosessions.GetSession` to check for errors?

---

I think the GetSession interface should also return the error rather than returning nil. 

```golang
func GetSession(c echo.Context) (*sessions.Session, err) {
	session, ok := c.Get("session").(*sessions.Session)
	if !ok {
		return nil, "Failed getting session."
	}
	return session, nil
}
```

---

The session will never be nil because it is being set in the sessions middleware. In `ginsessions` also it can return nil but it is left unhandled. Ideally we should move to a more robust package like  https://github.com/gorilla/sessions but didn't want to combine many major changes.

Refer:
https://github.com/kinvolk/nebraska/blob/7ecb9fcc6c1ff7f8358da5c04308b67b8eb9587f/backend/pkg/sessions/gin/gin.go#L24
https://github.com/kinvolk/nebraska/blob/7ecb9fcc6c1ff7f8358da5c04308b67b8eb9587f/backend/pkg/sessions/context.go#L16

---

Thanks for the explanation.